x-pack/plugin/repository-encrypted/src/test/java/org/elasticsearch/repositories/encrypted/ChainingInputStreamTests.java

```diff
@@ -0,0 +1,573 @@
+        test = new ChainingInputStream() {
+            boolean second = false;
+            @Override
+            InputStream nextComponent(InputStream currentComponentIn) throws IOException {
+                if (false == second) {
+                    second = true;
+                    return new ByteArrayInputStream(b2);
+                } else {
+                    return null;
+                }
+            }
+        };
+        test.currentIn = new ByteArrayInputStream(b1);
+        long skipArg = b1.length + 1 + Randomness.get().nextInt(b2.length);
+        nSkip = test.skip(skipArg);
+        assertThat(nSkip, Matchers.is(skipArg));
+        byte[] rest = test.readAllBytes();
+        assertThat((long)rest.length, Matchers.is(b1.length + b2.length - nSkip));
+        for (int i = rest.length - 1; i >= 0; i--) {
+            assertThat(rest[i], Matchers.is(b2[i + (int)nSkip - b1.length]));
+        }
+    }
+
+    public void testEmptyChain() throws Exception {
+        ChainingInputStream emptyStream = newEmptyStream();
+        assertThat(emptyStream.read(), Matchers.is(-1));
+        emptyStream = newEmptyStream();
+        byte[] b = new byte[1 + Randomness.get().nextInt(8)];
+        int off = Randomness.get().nextInt(b.length);
+        assertThat(emptyStream.read(b, off, b.length - off), Matchers.is(-1));
+        emptyStream = newEmptyStream();
+        assertThat(emptyStream.available(), Matchers.is(0));
+        emptyStream = newEmptyStream();
+        assertThat(emptyStream.skip(1 + Randomness.get().nextInt(32)), Matchers.is(0L));
+    }
+
+    public void testClose() throws Exception {
+        ChainingInputStream test1 = newEmptyStream();
+        test1.close();
+        IOException e = expectThrows(IOException.class, () -> {
+            test1.read();
+        });
+        assertThat(e.getMessage(), Matchers.is("Stream is closed"));
+        ChainingInputStream test2 = newEmptyStream();
+        test2.close();
+        byte[] b = new byte[2 + Randomness.get().nextInt(8)];
+        int off = Randomness.get().nextInt(b.length - 1);
+        e = expectThrows(IOException.class, () -> {
+            test2.read(b, off, Randomness.get().nextInt(b.length - off));
+        });
+        assertThat(e.getMessage(), Matchers.is("Stream is closed"));
+        ChainingInputStream test3 = newEmptyStream();
+        test3.close();
+        e = expectThrows(IOException.class, () -> {
+            test3.skip(Randomness.get().nextInt(32));
+        });
+        ChainingInputStream test4 = newEmptyStream();
+        test4.close();
+        e = expectThrows(IOException.class, () -> {
+            test4.available();
+        });
+        ChainingInputStream test5 = newEmptyStream();
+        test5.close();
+        e = expectThrows(IOException.class, () -> {
+            test5.reset();
+        });
+        ChainingInputStream test6 = newEmptyStream();
+        test6.close();
+        test6.mark(Randomness.get().nextInt());
```

Comment:
Can we make this a bit more explicit - it looks like it's just "tacked on" but I assume you're testing that `mark`  on a closed stream doesn't throw.
```suggestion
        try {
            test6.mark(Randomness.get().nextInt());
        } catch (Exception e) {
            assumeNoException("mark on a closed stream should succeed", e);
       }
```

Reply:
Yes, good point!